File: src/modlunky2/category/chain/cosmic.py
from typing import Set

from modlunky2.category.chain.common import (
    ChainMixin,
    ChainStepper,
)
from modlunky2.mem.entities import EntityType
from modlunky2.mem.state import Screen, State, WinState


class CosmicOceanChain(ChainMixin):
    @classmethod
    def make_stepper(cls) -> ChainStepper:
        return ChainStepper(cls.__name__, cls().collect_bow)

    def collect_bow(self, game_state: State, player_item_types: Set[EntityType]):
        if EntityType.ITEM_HOUYIBOW in player_item_types:
            return self.in_progress(self.carry_bow_to_hundun)

        if game_state.world > 2:
            return self.failed()

        return self.unstarted()

    def carry_bow_to_hundun(
        self, game_state: State, player_item_types: Set[EntityType]
    ):
        if game_state.win_state is not WinState.NO_WIN:
            return self.failed()

        world_level = (game_state.world, game_state.level)
        if world_level > (7, 3):
            return self.in_progress(self.win_via_co)

        if game_state.screen is not Screen.LEVEL_TRANSITION:
            return self.in_progress(self.carry_bow_to_hundun)

        if (
            EntityType.ITEM_HOUYIBOW in player_item_types
        ) or self.some_companion_has_item(game_state, EntityType.ITEM_HOUYIBOW):
            return self.in_progress(self.carry_bow_to_hundun)

        if world_level < (7, 1) and (
            EntityType.ITEM_HOUYIBOW in game_state.waddler_storage
        ):
            return self.in_progress(self.carry_bow_to_hundun)

        return self.failed()

    def win_via_co(self, game_state: State, _: Set[EntityType]):
        if game_state.win_state in (WinState.NO_WIN, WinState.COSMIC_OCEAN):
            return self.in_progress(self.win_via_co)

        return self.failed()

File: crates/ml2_trackers/src/chain_impl/cosmic.rs
//! Cosmic Ocean chain.
//!
//! A short three-step FSM: collect the Hou Yi bow -> keep carrying it
//! all the way to Hundun -> stay alive through the CO stack until a
//! CO win. `chain_impl/inputs.rs` precomputes the companion + waddler
//! queries so every step is a pure function over `ChainInputs`.

use crate::chain::{ChainStepResult, ChainStepper};
use crate::chain_impl::inputs::ChainInputs;
use crate::entity_types as et;
use crate::enums::{Screen, WinState};

pub fn make_stepper() -> ChainStepper<ChainInputs> {
    ChainStepper::new("CosmicOceanChain", collect_bow)
}

pub(crate) fn collect_bow(ctx: &ChainInputs) -> ChainStepResult<ChainInputs> {
    if ctx.player_items.contains(&et::ITEM_HOUYIBOW) {
        return ChainStepResult::in_progress(carry_bow_to_hundun);
    }
    // Bow lives in the jungle / volcana section (world 2). Past that
    // without picking it up = failure.
    if ctx.state.world > 2 {
        return ChainStepResult::failed();
    }
    ChainStepResult::unstarted()
}

pub(crate) fn carry_bow_to_hundun(ctx: &ChainInputs) -> ChainStepResult<ChainInputs> {
    if ctx.state.win_state != WinState::NoWin {
        return ChainStepResult::failed();
    }
    let wl = ctx.state.world_level();
    if wl > (7, 3) {
        return ChainStepResult::in_progress(win_via_co);
    }
    if ctx.state.screen != Screen::LevelTransition {
        return ChainStepResult::in_progress(carry_bow_to_hundun);
    }
    if ctx.player_items.contains(&et::ITEM_HOUYIBOW)
        || ctx.some_companion_has_item(et::ITEM_HOUYIBOW)
    {
        return ChainStepResult::in_progress(carry_bow_to_hundun);
    }
    // Bow can be safely stashed in Waddler's storage before world 7
    // and still counts as carried.
    if wl < (7, 1) && ctx.waddler_contains(et::ITEM_HOUYIBOW) {
        return ChainStepResult::in_progress(carry_bow_to_hundun);
    }
    ChainStepResult::failed()
}

pub(crate) fn win_via_co(ctx: &ChainInputs) -> ChainStepResult<ChainInputs> {
    // Stay in-progress through the CO climb until either the CO win
    // lands or a non-CO win-state comes through (means a different
    // ending fired).
    let ws = ctx.state.win_state;
    if ws == WinState::NoWin || ws == WinState::CosmicOcean {
        return ChainStepResult::in_progress(win_via_co);
    }
    ChainStepResult::failed()
}
